packages/eslint-plugin/src/rules/no-type-only-packages.ts

```diff
@@ -32,7 +32,11 @@ const rule = createRule({
     let hasNonEmptyFile = false;
 
     function isOutsidePackage(fileName: string): boolean {
-      return fileName.includes("node_modules") || path.relative(pkg!.dir, fileName).startsWith("..");
+      if (fileName.includes("node_modules") || path.relative(pkg!.dir, fileName).startsWith("..")) {
+        return true;
+      }
+      const other = findTypesPackage(fileName);
+      return !other || other.dir !== pkg!.dir;
```

Comment:
It turns out that this test explicitly includes files from other packages (including versioned subdirs) to test the rule, but we weren't actually checking this here. It's not clear to me why this wasn't breaking before, though.

Reply:
It's technically not the best to not cache this result, but since linting can be in an editor, we don't want to store a global cache or something. It shouldn't really cost that much.